replacement_scan.go

```diff
@@ -0,0 +1,66 @@
+package duckdb
+
+/*
+   #include <stdlib.h>
+   #include <duckdb.h>
+   void replacement_scan_cb(duckdb_replacement_scan_info info, const char *table_name, void *data);
+   typedef const char cchar_t;
+*/
+import "C"
+import (
+	"sync"
+	"unsafe"
+)
+
+type ReplacementScanCallback func(tableName string) (string, []any, error)
+
+var replacementScanState struct {
+	mtx      sync.Mutex
+	callback ReplacementScanCallback
+}
+
+func RegisterReplacementScan(connector *Connector, cb ReplacementScanCallback) {
+	replacementScanState.mtx.Lock()
+	replacementScanState.callback = cb
+	replacementScanState.mtx.Unlock()
+	C.duckdb_add_replacement_scan(connector.db, C.duckdb_replacement_callback_t(C.replacement_scan_cb), nil, C.duckdb_delete_callback_t(C.free))
+}
+
+//export replacement_scan_cb
+func replacement_scan_cb(info C.duckdb_replacement_scan_info, table_name *C.cchar_t, data *C.void) {
+	replacementScanState.mtx.Lock()
+	scanner := replacementScanState.callback
+	replacementScanState.mtx.Unlock()
+
+	tFunc, params, err := scanner(C.GoString(table_name))
+	if err != nil {
+		errstr := C.CString(err.Error())
+		C.duckdb_replacement_scan_set_error(info, errstr)
+		C.free(unsafe.Pointer(errstr))
+		return
+	}
+
+	fNameStr := C.CString(tFunc)
+	C.duckdb_replacement_scan_set_function_name(info, fNameStr)
+	defer C.free(unsafe.Pointer(fNameStr))
+
+	for _, v := range params {
+		switch x := v.(type) {
+		case string:
+			str := C.CString(x)
+			val := C.duckdb_create_varchar(str)
+			C.duckdb_replacement_scan_add_parameter(info, val)
+			C.free(unsafe.Pointer(str))
+			C.free(unsafe.Pointer(val))
```

Comment:
This should be `C.duckdb_destroy_value(&val)`, I think...

Reply:
Right, nice catch!